projects/graph/social/social.py

```diff
@@ -50,28 +51,83 @@ def populateGraph(self, numUsers, avgFriendships):
             self.addUser(f"User: {i}")
 
         # Create friendships
+        #itertools uses nested for loop under hood so this is inefficient at n^2
         possible_friendships = list(combinations(range(1, numUsers+1), 2))
         random.shuffle(possible_friendships)
-        friendships_on_avg = (possible_friendships * avgFriendships) // 2
-        actual_friendships = possible_friendships[:friendships_on_avg]
+        total_friendships = (possible_friendships * avgFriendships) // 2
+        actual_friendships = possible_friendships[:total_friendships]
         for friendship in actual_friendships:
             self.addFriendship(friendship[0], friendship[1])
+        #could also practice shuffling with fisher-yates
+
+        #more efficient refactor that is O(n)
+        # total_friendships = (possible_friendships * avgFriendships) // 2
+        # num_created = 0
+        # num_warnings = 0
+        # while num_created < total_friendships:
+        #     friendship = (random.randint(1, numUsers), random.randint(1, numUsers))
+        #     if self.addFriendship(friendship[0], friendship[1]):
+        #         num_created += 1
+        #     else: 
+        #         num_warnings += 1
+        # print("linear warning: {num_warnings}")
 
     def getAllSocialPaths(self, userID):
-        
         # Takes a user's userID as an argument
         # Returns a dictionary containing every user in that user's
         # extended network with the shortest friendship path between them.
         # The key is the friend's ID and the value is the path.
         
         visited = {}  # Note that this is a dictionary, not a set
-        # !!!! IMPLEMENT ME
+            if 
+
         return visited
+    
+    #bfs returns shortest path v dfs
+    def social_paths_bfs(self, starting_vertex, target):
+        q = deque()
+        visited = []
```

Comment:
great implementation of your shortest path!, if you can try to reduce the number of loops as this can solution can be optimized